ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/video.js

```diff
@@ -0,0 +1,29 @@
+// LOGIC: Extract the domain name from the URL(attr: href of link_open)
+// & UUID of the youtube video and render iframe is domain is youtube
+// and the given youtube link has a valid UUID
+
+module.exports = function installRule(markdownIt) {
+
+	const defaultRender = markdownIt.renderer.rules.link_open;
+	const videoIDRegex = /^.*(youtu\.be\/|v\/|u\/\w\/|embed\/|watch\?v=|&v=)([^#&?]*).*/;
+	const key = /.*(youtube|youtu.be).*/im;
+	// videoIDRegex extracts the unique 11 alpha-numeric code which is unique to every youtube video
+
+	markdownIt.renderer.rules.link_open = (tokens, idx, options, env, self) => {
+		const token = tokens[idx];
+		const link = token.attrGet('href');
+		const match = key.test(link);
+		const videoID = link.match(videoIDRegex);
+
+		if (!match || !videoID) return defaultRender(tokens, idx, options, env, self);
+
+		if (match && videoID[2].length == 11) {
+			const embedStart = '<div><iframe  width="560" height="315" src="https://www.youtube.com/embed/';
+			const embedEnd = '"  frameborder="0" allowfullscreen allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"></iframe></div>';
+			return embedStart + videoID[2] + embedEnd;
```

Comment:
Please use a template literal to format and indent the HTML string - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

Any reason why you chose width 560? I guess it makes sense to set a min and max width, but otherwise could you make it shrink and grow with the container?

Could you document these properties and are they necessary? `accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture` For example, for me the video doesn't even auto-play (and I don't think it should) so the autoplay property maybe is not needed. I don't know about the others.

Reply:
> Any reason why you chose width 560? I guess it makes sense to set a min and max width, but otherwise could you make it shrink and grow with the container?


`width ="560" and height="315"` is to maintain the aspect ratio 16:9 which is the default aspect ratio for the majority of videos and is used to prevent black bars,


**Autoplay** means that once you finish the video should the next one autoplay(_**auto-playing** the video in youtube recommendation queue, not the one embedded in the player_)

**Picture-in-picture** shrinks the video into a small player that you can move around your home screen and position over other apps. You can continue watching YouTube videos while using other apps on your mobile device with this feature turned on.


**accelerometer and gyroscope** 
This usually refers to when a user rotates or moves their device around, so I'm guessing this would be for when someone embeds a Virtual Reality or 360degree video?

**Encrypted media** 
Encrypted Media Extensions provides an API that enables web applications to interact with content protection systems, to allow playback of encrypted audio and video

Reply:
> if we would use max-width, min-width we might have black bars in the video player because of improper aspect ratio,

Why? The height should also grow and shrink with the width, thus the ratio is not affected. Also "Any reason why you chose width 560?"

Regarding the extra property, I get the feeling we could remove them all and it wouldn't change much, or am I missing something? I guess you copied that from somewhere else, which is fine, but in our case if we don't need them, we can remove them.

Reply:
>Regarding the extra property, I get the feeling we could remove them all and it wouldn't change much, or am I missing something? I guess you copied that from somewhere else, which is fine, but in our case if we don't need them, we can remove them.

Yes it is written on the template youtube provides for embedding videos


<img width="1127" alt="Annotation 2020-03-24 164157" src="https://user-images.githubusercontent.com/54576074/77419507-8937c900-6dee-11ea-9907-40a06362ae27.png">


``` htm

<iframe width="560" height="315" src="https://www.youtube.com/embed/vdHJjbHwR38" 
frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in
-picture" allowfullscreen></iframe>

```

Reply:
> Why? The height should also grow and shrink with the width, thus the ratio is not affected. Also "Any reason why you chose width 560?"

Okay, dimension 560X315 =~ 16:9 (which is desired since most videos are in that aspect ratio) 

we can change it to width to max-width in iframe and calculate the height so it is appx 16:9,
I guess that would be a better implementation, what do you think?


#### [Current 560X315]
![Screenshot (10)](https://user-images.githubusercontent.com/54576074/77424082-8ccf4e00-6df6-11ea-837c-29e257de6401.png)

#### Changing width to 640~max width lead to black bars in video if we do not correct the aspect ratio in iframe
![Screenshot (11)](https://user-images.githubusercontent.com/54576074/77424094-8fca3e80-6df6-11ea-9336-0e54009dcb5f.png)


####  Correcting the aspect ratio
![Screenshot (12)](https://user-images.githubusercontent.com/54576074/77424126-9953a680-6df6-11ea-8d69-ba2487f6437f.png)

Reply:
> it is written on the template youtube provides for embedding videos, so no additional reason for using width 560 besides the fact it was provided by youtube when I checked the specification for the embedded link,

That width is determined by your browser's window size AFAIK. You could try [a CSS trick](https://stackoverflow.com/a/35153397) to keep aspect ratio of iframe instead of specifying arbitrary numbers.